packages/wasm/crate/src/planner.rs

```diff
@@ -486,111 +463,48 @@ pub async fn plan_transaction(
 
     let mut iterations = 0usize;
 
-    while let Some(required) = actions
-        .balance_with_fee_estimate(&gas_prices, &fee_tier)
-        .required()
-        .next()
-    {
-        // Spend a single note towards the required balance, if possible.
-        let Some(note) = notes_by_asset_id
+    // Now iterate over the action list's imbalances to balance the transaction.
+    while let Some(required) = actions_list.balance_with_fee().required().next() {
+        // Find a single note to spend towards the required balance.
+        let note = notes_by_asset_id
             .get_mut(&required.asset_id)
-            .expect("we already queried")
+            .expect("we already made a notes request for each required asset")
             .pop()
-        else {
-            return Err(anyhow!(
-                "ran out of notes to spend while planning transaction, need {} of asset {}",
-                required.amount,
-                required.asset_id,
-            )
-            .into());
-        };
-        actions.push(SpendPlan::new(&mut OsRng, note.note, note.position).into());
+            .ok_or_else(|| {
+                anyhow!(
+                    "ran out of notes to spend while planning transaction, need {} of asset {}",
+                    required.amount,
+                    required.asset_id,
+                )
+            })?;
 
-        // Recompute the change outputs, without accounting for fees.
-        actions.refresh_change(change_address.clone());
-        // Now re-estimate the fee of the updated transaction and adjust the change if possible.
-        let fee = actions.fee_estimate(&gas_prices, &fee_tier);
-        actions.adjust_change_for_fee(fee);
+        // Add a spend for that note to the action list.
+        actions_list.push(SpendPlan::new(&mut OsRng, note.note, note.position));
+
+        // Refresh the fee estimate and change outputs.
+        actions_list.refresh_fee_and_change(OsRng, &gas_prices, &fee_tier, &change_address);
 
         iterations += 1;
         if iterations > 100 {
             return Err(anyhow!("failed to plan transaction after 100 iterations").into());
         }
     }
 
-    let fee = actions.fee_estimate(&gas_prices, &fee_tier);
-
-    let mut plan = TransactionPlan {
-        actions: actions
-            .actions
-            .into_iter()
-            .chain(actions.change_outputs.into_values().map(Into::into))
-            .collect(),
-        transaction_parameters: TransactionParameters {
-            expiry_height: request.expiry_height,
-            chain_id,
-            fee,
-        },
-        detection_data: None,
-        memo: None,
-    };
-
+    // Add memo to the transaction plan.
+    let mut memo: Option<MemoPlan> = None;
```

Comment:
Totally optional but you could consider setting the memo like so instead of using a mut:

```rust
    let memo = if let Some(pb_memo_plaintext) = request.memo {
        Some(MemoPlan::new(&mut OsRng, pb_memo_plaintext.try_into()?))
    } else if actions_list.requires_memo() {
        let plaintext = MemoPlaintext::new(change_address, String::new())?;
        Some(MemoPlan::new(&mut OsRng, plaintext))
    } else {
        None
    };
``

Reply:
changed